src/python/pants/backend/terraform/dependency_inference.py

```diff
@@ -0,0 +1,186 @@
+
+
+    paths = set()
+    for filename in sys.argv[1:]:
+        with open(filename, "rb") as f:
+            content = f.read()
+        paths |= extract_module_source_paths(PurePath(filename).parent, content)
+
+    for path in paths:
+        print(path)
+    """
+    ).encode("utf-8"),
+    is_executable=True,
+)
+
+
+class PythonHcl2Subsystem(PythonToolRequirementsBase):
+    options_scope = "python-hcl2"
+    help = "Used to parse Terraform modules to infer their dependencies."
+
+    default_version = "python-hcl2==3.0.1"
+
+    register_interpreter_constraints = True
+    default_interpreter_constraints = ["CPython>=3.7"]
+
+
+@dataclass(frozen=True)
+class ParserSetup:
+    pex: VenvPex
+
+
+@rule
+async def setup_parser(hcl2_subsystem: PythonHcl2Subsystem) -> ParserSetup:
+    parser_digest = await Get(Digest, CreateDigest([PARSER]))
+
+    parser_pex = await Get(
+        VenvPex,
+        PexRequest(
+            output_filename="tf_parser.pex",
+            internal_only=True,
+            requirements=hcl2_subsystem.pex_requirements(),
+            interpreter_constraints=hcl2_subsystem.interpreter_constraints,
+            main=EntryPoint(PurePath(PARSER.path).stem),
+            sources=parser_digest,
+        ),
+    )
+
+    return ParserSetup(parser_pex)
+
+
+@dataclass(frozen=True)
+class ParseTerraformModuleSources:
+    sources_digest: Digest
+    paths: tuple[str, ...]
+
+
+@rule
+async def setup_process_for_parse_terraform_module_sources(
+    request: ParseTerraformModuleSources, parser: ParserSetup
+) -> Process:
+    process = await Get(
+        Process,
+        VenvPexProcess(
+            parser.pex,
+            argv=request.paths,
+            input_digest=request.sources_digest,
+            description="Parse Terraform module sources.",
+        ),
+    )
+    return process
+
+
+class InferTerraformModuleDependenciesRequest(InferDependenciesRequest):
+    infer_from = TerraformModuleSources
+
+
+@rule
+async def infer_terraform_module_dependencies(
+    request: InferTerraformModuleDependenciesRequest,
+) -> InferredDependencies:
+    hydrated_sources = await Get(HydratedSources, HydrateSourcesRequest(request.sources_field))
+
+    paths = OrderedSet(
+        filename for filename in hydrated_sources.snapshot.files if filename.endswith(".tf")
+    )
+    result = await Get(
+        ProcessResult,
+        ParseTerraformModuleSources(
+            sources_digest=hydrated_sources.snapshot.digest,
+            paths=tuple(paths),
+        ),
+    )
+    candidate_spec_paths = [line for line in result.stdout.decode("utf-8").split("\n") if line]
+
+    # For each path, see if there is a `terraform_module` target at the specified spec_path.
+    candidate_targets = await Get(
+        Targets, AddressSpecs([MaybeEmptySiblingAddresses(path) for path in candidate_spec_paths])
+    )
+    terraform_module_addresses = [
+        tgt.address for tgt in candidate_targets if tgt.has_field(TerraformModuleSources)
+    ]
```

Comment:
This needs to use the generic mechanism for ambiguous dependencies - unless we enforce our ban that there's only one `terraform_module` per dir (https://github.com/pantsbuild/pants/issues/12921). 

I'm fine punting on this, but please add a TODO.

Reply:
Bump on a TODO please

Reply:
Done.